src/System.Text.Json/src/System/Text/Json/Serialization/MemberAccessor.cs

```diff
@@ -2,92 +2,64 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
+using System.Diagnostics;
 using System.Reflection;
-using System.Runtime.CompilerServices;
 
 namespace System.Text.Json
 {
-    internal static class MemberAccessor
+    internal abstract class MemberAccessor
     {
-        private delegate TProperty GetProperty<TClass, TProperty>(TClass obj);
-        private delegate TProperty GetPropertyByRef<TClass, TProperty>(ref TClass obj);
+        public abstract JsonClassInfo.ConstructorDelegate CreateConstructor(Type classType);
 
-        private delegate void SetProperty<TClass, TProperty>(TClass obj, TProperty value);
-        private delegate void SetPropertyByRef<TClass, TProperty>(ref TClass obj, TProperty value);
+        public abstract object ImmutableCollectionCreateRange(Type constructingType, Type elementType);
 
-        private delegate Func<object, TProperty> GetPropertyByRefFactory<TClass, TProperty>(GetPropertyByRef<TClass, TProperty> set);
-        private delegate Action<object, TProperty> SetPropertyByRefFactory<TClass, TProperty>(SetPropertyByRef<TClass, TProperty> set);
+        public abstract object ImmutableDictionaryCreateRange(Type constructingType, Type elementType);
 
-        private static readonly MethodInfo s_createStructPropertyGetterMethod = new GetPropertyByRefFactory<int, int>(CreateStructPropertyGetter)
-            .Method.GetGenericMethodDefinition();
-
-        private static readonly MethodInfo s_createStructPropertySetterMethod = new SetPropertyByRefFactory<int, int>(CreateStructPropertySetter)
-            .Method.GetGenericMethodDefinition();
-
-        internal static Func<object, TProperty> CreatePropertyGetter<TClass, TProperty>(PropertyInfo propertyInfo)
+        protected MethodInfo ImmutableCollectionCreateRangeMethod(Type constructingType, Type elementType)
         {
-            MethodInfo getMethodInfo = propertyInfo.GetGetMethod();
-
-            if (typeof(TClass).IsValueType)
-            {
-                var factory = CreateDelegate<GetPropertyByRefFactory<TClass, TProperty>>(s_createStructPropertyGetterMethod.MakeGenericMethod(typeof(TClass), typeof(TProperty)));
-                var propertyGetter = CreateDelegate<GetPropertyByRef<TClass, TProperty>>(getMethodInfo);
+            MethodInfo createRangeMethod = FindImmutableCreateRangeMethod(constructingType);
 
-                return factory(propertyGetter);
-            }
-            else
+            if (createRangeMethod == null)
```

Comment:
@YohDeadfall - we are [annotating `System.Text.Json` with nullability](https://github.com/dotnet/runtime/pull/528) and it looks like this branch is not covered by any tests . When can `createRangeMethod` be null? And if this is a valid branch, we should add a test. Using `Debug.Fail` doesn't work in release.

Reply:
It's not mine code, but @layomia's (: [The comment](https://github.com/dotnet/runtime/blob/4f9ae42d861fcb4be2fcd5d3d55d5f227d30e723/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/MemberAccessor.cs#L56-L60) says that this shouldn't happen. Probably it would be better to throw exception directly from `FindImmutableCreateRangeMethod`.

Reply:
> It's not mine code

Yes, just realized that when looking at the commit/history more closely (this PR moved the code). Sorry about the ping :)

This is the PR I am interested in: https://github.com/dotnet/corefx/pull/37710

Reply:
No worries, feel free to ping me for any question (:

By the way, is it the right time to move the field support to the new repo? Have you finished optimizations?